beetmoverscript/zip.py

```diff
@@ -0,0 +1,206 @@
+    }
+
+
+def _ensure_files_in_archive_have_decent_sizes(zip_path, zip_metadata, zip_max_size_in_mb):
+    for file_name, file_metadata in zip_metadata.items():
+        compressed_size = file_metadata['compress_size']
+        real_size = file_metadata['file_size']
+        compressed_size_size_in_mb = compressed_size // (1024 * 1024)
+
+        if compressed_size_size_in_mb > zip_max_size_in_mb:
+            raise TaskVerificationError(
+                'In archive "{}", compressed file "{}" is too big. Max accepted size (in MB): {}. File size (in MB): {}'.format(
+                    zip_path, file_name, zip_max_size_in_mb, compressed_size_size_in_mb
+                )
+            )
+
+        compression_ratio = real_size / compressed_size
+        if compression_ratio > ZIP_MAX_COMPRESSION_RATIO:
+            raise TaskVerificationError(
+                'In archive "{}", file "{}" has a suspicious compression ratio. Max accepted: {}. Found: {}'.format(
+                    zip_path, file_name, ZIP_MAX_COMPRESSION_RATIO, compression_ratio
+                )
+            )
+
+    log.info('Archive "{}" contains files with legitimate sizes.'.format(zip_path))
+
+
+def _ensure_all_expected_files_are_present_in_archive(zip_path, files_in_archive, expected_files):
+    files_in_archive = set(files_in_archive)
+
+    unique_expected_files = set(expected_files)
+    if len(expected_files) != len(unique_expected_files):
+        duplicated_files = [file for file in unique_expected_files if expected_files.count(file) > 1]
+        raise TaskVerificationError(
+            'Found duplicated expected files in archive "{}": {}'.format(zip_path, duplicated_files)
+        )
+
+    for file in files_in_archive:
+        if os.path.isabs(file):
+            raise TaskVerificationError(
+                'File "{}" in archive "{}" cannot be an absolute one.'.format(file, zip_path)
+            )
+        if os.path.normpath(file) != file:
+            raise TaskVerificationError(
+                'File "{}" in archive "{}" cannot contain up-level reference nor redundant separators'.format(
+                    file, zip_path
+                )
+            )
+        if file not in unique_expected_files:
+            raise TaskVerificationError(
+                'File "{}" present in archive "{}" is not expected. Expected: {}'.format(
+                    file, zip_path, unique_expected_files
+                )
+            )
+
+    if len(files_in_archive) != len(unique_expected_files):
+        missing_expected_files = [file for file in unique_expected_files if file not in files_in_archive]
+        raise TaskVerificationError(
+            'Expected files are missing in archive "{}": {}'.format(zip_path, missing_expected_files)
+        )
+
+    log.info('Archive "{}" contains all expected files: {}'.format(zip_path, unique_expected_files))
+
+
+def _extract_and_check_output_files(zip_file, expected_files_in_archive):
```

Comment:
Similarly as one of the functions above, I'd rather keep the `s/expected_files_in_arhive/zip_metadata` and do that `zip_metadata.keys()` processing in this function to enhance the logic reading. But it's a matter of preference, I don't have a strong feeling nor way or another.

Reply:
Done in https://github.com/mozilla-releng/beetmoverscript/pull/163/commits/86f9770b59d509a3ad11871354248b165b9122c3